File: jFamilyCounselorImpl/src/main/scala/ro/lrg/jfamilycounselor/util/parse/visitor/MethodDeclarationVisitor.scala
package ro.lrg.jfamilycounselor.util.parse.visitor

import org.eclipse.jdt.core.IMethod
import org.eclipse.jdt.core.dom.MethodDeclaration
import ro.lrg.jfamilycounselor.util.parse.MemberResolvingVisitor

class MethodDeclarationVisitor(method: IMethod)
    extends MemberResolvingVisitor[MethodDeclaration](method) {
  var result: MethodDeclaration = _

  override def visit(node: MethodDeclaration): Boolean = {
    if (member == node.resolveBinding.getJavaElement)
      result = node
    true
  }

}

File: jfamilycounselor/src/ro/lrg/jfamilycounselor/capability/generic/parse/MethodDeclarationVisitor.java
package ro.lrg.jfamilycounselor.capability.generic.parse;

import java.util.Optional;

import org.eclipse.jdt.core.IMethod;
import org.eclipse.jdt.core.dom.ASTVisitor;
import org.eclipse.jdt.core.dom.MethodDeclaration;

class MethodDeclarationVisitor extends ASTVisitor {
    private Optional<MethodDeclaration> lastNode = Optional.empty();
    private final IMethod iMethod;

    public MethodDeclarationVisitor(IMethod iMethod) {
	this.iMethod = iMethod;
    }

    public boolean visit(MethodDeclaration node) {
	
	if (Optional.ofNullable(node.resolveBinding()).map(b -> b.getJavaElement()).stream().anyMatch(j -> j.equals(iMethod))) {
	    lastNode = Optional.of(node);
	}

	return false;
    }

    public Optional<MethodDeclaration> getLastNode() {
	return lastNode;
    }
}
